pkg/runtime/bulk_subscriber.go

```diff
@@ -0,0 +1,706 @@
+			spans[n] = span
+			n++
+		}
+	}
+	spans = spans[:n]
+	defer endSpans(spans)
+	start := time.Now()
+	resp, err := a.appChannel.InvokeMethod(ctx, req)
+	elapsed := diag.ElapsedSince(start)
+
+	if err != nil {
+		bulkSubDiag.statusWiseDiag[string(pubsub.Retry)] += int64(len(msg.entries))
+		bulkSubDiag.elapsed = elapsed
+		populateBulkSubscribeResponsesWithError(msg.entries, bulkResponses, &entryIDIndexMap, err)
+		return errors.Wrap(err, "error from app channel while sending pub/sub event to app")
+	}
+
+	statusCode := int(resp.Status().Code)
+
+	for _, span := range spans {
+		if span != nil {
+			m := diag.ConstructSubscriptionSpanAttributes(msg.topic)
+			diag.AddAttributesToSpan(span, m)
+			diag.UpdateSpanStatusFromHTTPStatus(span, statusCode)
+		}
+	}
+
+	_, body := resp.RawData()
+
+	if (statusCode >= 200) && (statusCode <= 299) {
+		// Any 2xx is considered a success.
+		var appBulkResponse pubsub.AppBulkResponse
+		err := json.Unmarshal(body, &appBulkResponse)
+		if err != nil {
+			bulkSubDiag.statusWiseDiag[string(pubsub.Success)] += int64(len(msg.entries))
+			bulkSubDiag.elapsed = elapsed
+			populateBulkSubscribeResponsesWithError(msg.entries, bulkResponses, &entryIDIndexMap, err)
+			return errors.Wrap(err, "failed unmarshalling app response for bulk subscribe")
+		}
+
+		var hasAnyError bool
+		for _, response := range appBulkResponse.AppResponses {
+			if entryID, ok := entryIDIndexMap[response.EntryID]; ok {
+				switch response.Status {
+				case "":
+					// When statusCode 2xx, Consider empty status field OR not receiving status for an item as retry
+					fallthrough
+				case pubsub.Retry:
+					bulkSubDiag.statusWiseDiag[string(pubsub.Retry)]++
+					(*bulkResponses)[entryID].EntryID = response.EntryID
+					(*bulkResponses)[entryID].Error = errors.Errorf("RETRY required while processing bulk subscribe event for entry id: %v", response.EntryID)
+					hasAnyError = true
+				case pubsub.Success:
+					bulkSubDiag.statusWiseDiag[string(pubsub.Success)]++
+					(*bulkResponses)[entryID].EntryID = response.EntryID
+					(*bulkResponses)[entryID].Error = nil
+				case pubsub.Drop:
+					bulkSubDiag.statusWiseDiag[string(pubsub.Drop)]++
+					log.Warnf("DROP status returned from app while processing pub/sub event %v", response.EntryID)
+					(*bulkResponses)[entryID].EntryID = response.EntryID
+					(*bulkResponses)[entryID].Error = nil
+				default:
+					// Consider unknown status field as error and retry
+					bulkSubDiag.statusWiseDiag[string(pubsub.Retry)]++
+					(*bulkResponses)[entryID].EntryID = response.EntryID
+					(*bulkResponses)[entryID].Error = errors.Errorf("unknown status returned from app while processing bulk subscribe event %v: %v", response.EntryID, response.Status)
+					hasAnyError = true
+				}
+			} else {
+				log.Warnf("Invalid entry id received from app while processing pub/sub event %v", response.EntryID)
+				continue
+			}
+		}
+		for _, item := range msg.entries {
```

Comment:
Reviewers: This iteration is necessary to take care of cases where-in response for that entryID is not sent back by the App. Those messages have fall under retry category.